ui/components/HeadProvider/index.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useContext, createContext, ReactNode } from 'react'
+import { globalHistory } from '@reach/router'
+
+interface NewScriptTagParams {
+  type: 'script'
+  script: string
+  texttype: string
+}
+
+interface NewTitleTagParams {
+  type: 'title'
+  text: string
+}
+
+interface NewMetaTagParams {
+  type: 'meta'
+  name: 'description' | 'keywords' | 'author' | 'viewport'
+  content: string
+}
+
+type NewTagParams = NewScriptTagParams | NewTitleTagParams | NewMetaTagParams
+
+export interface Head {
+  tags: JSX.Element[]
+  addTag: (params: NewTagParams) => void
+}
+
+const HeadContext = createContext<Head>({
+  tags: [],
+  addTag: () => {},
+})
+
+interface HeadProviderProps {
+  children: ReactNode
+  tags: JSX.Element[]
+}
+
+export function HeadProvider(props: HeadProviderProps) {
+  let { children, tags } = props
+  globalHistory.listen(() => {
+    if (typeof window !== 'undefined') {
+      const metaElements = Array.from(
+        document.getElementsByTagName('head')[0].getElementsByTagName('meta'),
+      )
+
+      const scriptElements = Array.from(
+        document.getElementsByTagName('head')[0].getElementsByTagName('script'),
+      )
+
+      metaElements.map((itm) => {
+        if (itm.parentNode) itm.parentNode.removeChild(itm)
+      })
+
+      scriptElements.map((itm) => {
+        if (itm.parentNode) itm.parentNode.removeChild(itm)
+      })
+    }
+  })
+  return (
+    <HeadContext.Provider
+      value={{
+        tags,
+        addTag: (params) => {
+          if (params.type === 'title') {
+            const current = tags.find((tag) => tag.type === 'title')
+            if (!current) tags.push(<title>{params.text}</title>)
+            else tags[tags.indexOf(current)] = <title>{params.text}</title>
+            if (
+              typeof window !== 'undefined' &&
+              document.getElementsByTagName('title')[0].text !== params.text
+            )
+              document.getElementsByTagName('title')[0].text = params.text
+          } else if (params.type === 'script') {
+            const current = tags.find(
+              (tag) => tag.type === 'script' && tag.type === params.texttype,
+            )
```

Comment:
Using `type` is the key for de-deplication will lead to scripts replacing each other unintentionally. Also, if a re-render happens that changes a script type, it won't know how to remove the old version.

Reply:
Any idea on how to dedup? I've considered creating a hash and comparing